Add DeepCopy to Unstructured

sttts opened the pull request:
Fixes https://github.com/kubernetes/kubernetes/issues/40657

sttts committed:
Add DeepCopy to Unstructured

sttts committed:
apimachinery: decompose meta/v1/unstructured in order to cut import cycle

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/40725)
<!-- Reviewable:end -->

sttts commented:
@k8s-bot unit test this

deads2k commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go:
```diff
@@ -80,6 +80,31 @@ func (obj *UnstructuredList) UnstructuredContent() map[string]interface{} {
+	}
+}
+
+func deepCopyJSON(x interface{}) interface{} {
```
This function assumes that we terminate in "simple" leaf node types, right?  Otherwise we'd end in a shallow copy.

I think the assumption holds true for all reasonable uses of unstructured, I just want to be sure I've understood correctly.

sttts replied:
This is my understanding of the `Unstructured` data type (which resembles JSON).

smarterclayton replied:
Nothing in unstructured should be in a leaf that isn't a string, nil, boolean, int, or float.  So this looks ok to me.

deads2k commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/zz_generated.deepcopy.go:
```diff
@@ -0,0 +1,62 @@
+	{
+		in := in.(*UnstructuredObjectConverter)
+		out := out.(*UnstructuredObjectConverter)
+		*out = *in
```
This looks off.  What is this this method supposed to do again?

smarterclayton replied:
I think this shouldn't have a deep copier.

deads2k commented:
a couple questions.  lgtm.

sttts commented:
The deepcopy funcs are not registered anywhere. Because unstructured also imports metav1, we cannot even piggyback on metav1 to get registered.

If we want this, we probably have to move it into metav1 itself. Anything that speaks against that?

ash2k commented:
It would be nice to have those deep copy functions registered in the schema (and also that means I can register them with my own instance of Scheme).

smarterclayton commented:
Moving unstructured?  Or just the deep copies?  I'm against the former,
very strongly.

On Thu, Mar 9, 2017 at 7:07 PM, Mikhail Mazurskiy <notifications@github.com>
wrote:

> It would be nice to have those deep copy functions registered in the
> schema (and also that means I can register them with my own instance of
> Scheme).
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/40725#issuecomment-285526591>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p-5i5Sh9MnPXjF2wff5Xyd6Ukpkcks5rkJQsgaJpZM4LyUeN>
> .
>

ash2k commented:
I only want:
- those deep copy functions to be useable from my code (outside of Kubernetes)
- be able to use an instance of `Scheme` to do `scheme.Copy(object_which_has_a_field_of_Unstructured_or_slice_of_Unstructured)`

How to achieve that you guys know much better than I :)

sttts commented:
@ash2k this will not be in 1.6 because it would need some rewrite/refactoring of the meta package in apimachinery.

Wouldn't you be able to check for the `Unstructured` type and then call something like `deepCopyJSON` of this PR in your code? At least for the time being.

ash2k commented:
@sttts No worries, looks like `Scheme.DeepCopy()` works pretty well without any copy functions so nothing is blocking me. Thanks!

smarterclayton commented:
Generic unstructured should be a subpackage of runtime.  We coupled a
versioned type (because of the accessors) to a generic concept.

I'd be supportive of that refactor happening st some point.

On Mar 10, 2017, at 6:12 AM, Mikhail Mazurskiy <notifications@github.com>
wrote:

@sttts <https://github.com/sttts> No worries, looks like Scheme.DeepCopy()
works pretty well without any copy functions so nothing is blocking me.
Thanks!

—
You are receiving this because you were assigned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/40725#issuecomment-285643230>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/ABG_p-t4aPIcd5VfRGF1BMHOkI6GjXwLks5rkTAYgaJpZM4LyUeN>
.

NickrenREN committed:
Add exponential backoff to openstack loadbalancer functions
Using  exponential backoff to lower openstack load and reduce API call throttling

ash2k commented:
Ping @sttts. How can I help? It's been 3 months since the PR was opened. Just don't want this work to go to waste.

sttts committed:
Fixup code using meta/v1/unstructured

sttts committed:
Update generated files

sttts commented:
@ash2k I have split the unstructured package into the acutal types and the helpers. The types can now be imported by metav1.

@smarterclayton where on a scale between -5 and 5 regarding ugliness is the new Accessor alias?

sttts commented:
This breaks of course all dynamic `ObjectMetaAccessor` type checks.

sttts commented:
Replaced by https://github.com/kubernetes/kubernetes/pull/48544.
